File: src/web/src/lib/types/claim.ts
/**
 * @fileoverview TypeScript types and interfaces for insurance claims
 * Implements HIPAA-compliant data structures for claims processing
 * @version 1.0.0
 */

// Internal imports
import { IUser } from '../types/user';
import { IHealthRecord } from '../types/healthRecord';

/**
 * Enum defining possible claim statuses with extended states
 */
export enum ClaimStatus {
    DRAFT = 'DRAFT',
    SUBMITTED = 'SUBMITTED',
    UNDER_REVIEW = 'UNDER_REVIEW',
    APPROVED = 'APPROVED',
    REJECTED = 'REJECTED',
    PENDING_INFO = 'PENDING_INFO',
    APPEALED = 'APPEALED'
}

/**
 * Enum defining comprehensive types of insurance claims
 */
export enum ClaimType {
    MEDICAL = 'MEDICAL',
    PHARMACY = 'PHARMACY',
    DENTAL = 'DENTAL',
    VISION = 'VISION',
    MENTAL_HEALTH = 'MENTAL_HEALTH',
    PREVENTIVE = 'PREVENTIVE'
}

/**
 * Interface for encryption metadata tracking
 */
interface IEncryptionMetadata {
    algorithm: string;
    keyId: string;
    initVector: string;
    lastEncryptedAt: Date;
}

/**
 * Interface for access logging
 */
interface IAccessLog {
    userId: string;
    accessedAt: Date;
    action: string;
    ipAddress: string;
}

/**
 * Interface for security metadata
 */
interface ISecurityMetadata {
    encryptionLevel: string;
    dataClassification: string;
    lastSecurityReview: Date;
    accessControlList: string[];
}

/**
 * Interface for compliance checks
 */
interface IComplianceCheck {
    type: string;
    status: boolean;
    checkedAt: Date;
    checkedBy: string;
    findings: string[];
}

/**
 * Interface for audit trail entries
 */
interface IAuditEntry {
    timestamp: Date;
    action: string;
    performedBy: string;
    details: string;
    systemMetadata: {
        ipAddress: string;
        userAgent: string;
    };
}

/**
 * Enhanced interface for claim supporting documents with security metadata
 */
export interface IClaimDocument {
    id: string;
    type: string;
    title: string;
    url: string;
    uploadedAt: Date;
    encryptionMetadata: IEncryptionMetadata;
    accessLog: IAccessLog[];
    hipaaCompliant: boolean;
}

/**
 * Enhanced main interface for insurance claim data structure
 * with security and compliance features
 */
export interface IClaim {
    id: string;
    version: number;
    claimNumber: string;
    patientId: string;
    providerId: string;
    type: ClaimType;
    serviceDate: Date;
    submissionDate: Date;
    status: ClaimStatus;
    amount: number;
    documents: IClaimDocument[];
    healthRecordId: string;
    auditTrail: IAuditEntry[];
    securityMetadata: ISecurityMetadata;
    complianceChecks: IComplianceCheck[];
}

/**
 * Enhanced interface for claim submission data with compliance requirements
 */
export interface IClaimSubmission {
    type: ClaimType;
    serviceDate: Date;
    providerId: string;
    amount: number;
    documents: File[];
    healthRecordId: string;
    consentAcknowledgment: boolean;
    hipaaAuthorization: boolean;
}

File: src/android/app/src/main/kotlin/com/austa/superapp/features/claims/domain/models/Claim.kt
package com.austa.superapp.features.claims.domain.models

import com.austa.superapp.core.constants.AppConstants
import kotlinx.serialization.Serializable // v1.5.1
import java.util.Date
import java.util.UUID

/**
 * Enhanced enum class defining possible claim statuses with validation logic
 */
@Serializable
enum class ClaimStatus {
    SUBMITTED,
    IN_REVIEW,
    APPROVED,
    REJECTED,
    PENDING_INFO;

    /**
     * Validates if status transition is allowed based on HIPAA compliance rules
     * @param newStatus The target status to transition to
     * @return Boolean indicating if transition is valid
     */
    fun isValidTransition(newStatus: ClaimStatus): Boolean {
        return when (this) {
            SUBMITTED -> setOf(IN_REVIEW, PENDING_INFO).contains(newStatus)
            IN_REVIEW -> setOf(APPROVED, REJECTED, PENDING_INFO).contains(newStatus)
            PENDING_INFO -> setOf(IN_REVIEW).contains(newStatus)
            APPROVED, REJECTED -> false
        }
    }
}

/**
 * Enhanced enum class defining types of insurance claims with metadata
 */
@Serializable
enum class ClaimType {
    MEDICAL,
    PHARMACY,
    DENTAL,
    VISION;

    /**
     * Determines if claim type requires pre-authorization based on policy rules
     * @return Boolean indicating pre-authorization requirement
     */
    fun requiresPreAuth(): Boolean {
        return when (this) {
            MEDICAL -> true
            PHARMACY -> false
            DENTAL -> true
            VISION -> false
        }
    }
}

/**
 * Enhanced data class for secure claim supporting documents with HIPAA compliance
 */
@Serializable
data class ClaimDocument(
    val id: String = UUID.randomUUID().toString(),
    val type: String,
    val title: String,
    val encryptedUrl: String,
    val uploadedAt: Date = Date(),
    val uploadedBy: String,
    val hashValue: String,
    val containsPHI: Boolean = false,
    val retentionDate: Date
)

/**
 * Enhanced data class for comprehensive claim metadata with security context
 */
@Serializable
data class ClaimMetadata(
    val policyNumber: String,
    val insuranceProvider: String,
    val facility: String,
    val diagnosisCodes: List<String>,
    val procedureCodes: List<String>,
    val coveragePercentage: Double,
    val deductibleApplied: Double,
    val preAuthNumber: String? = null,
    val securityLabels: Map<String, String> = mapOf(),
    val notes: String? = null
)

/**
 * Enhanced data class for secure audit trail entries
 */
@Serializable
data class AuditEntry(
    val timestamp: Date = Date(),
    val userId: String,
    val action: String,
    val details: String,
    val ipAddress: String,
    val deviceInfo: String
)

/**
 * Enhanced main data class representing a HIPAA-compliant insurance claim
 * with comprehensive security features and audit trails
 */
@Serializable
data class Claim(
    val id: String = UUID.randomUUID().toString(),
    val patientId: String,
    val providerId: String,
    val type: ClaimType,
    val status: ClaimStatus = ClaimStatus.SUBMITTED,
    val amount: Double,
    val serviceDate: Date,
    val submissionDate: Date = Date(),
    val healthRecordId: String,
    val documents: List<ClaimDocument> = listOf(),
    val metadata: ClaimMetadata,
    val auditTrail: List<AuditEntry> = listOf(),
    val securityContext: Map<String, String> = mapOf()
) {
    init {
        require(amount > 0 && amount <= AppConstants.CLAIMS.MAX_CLAIM_AMOUNT) {
            "Claim amount must be between 0 and ${AppConstants.CLAIMS.MAX_CLAIM_AMOUNT}"
        }
        require(documents.size <= AppConstants.CLAIMS.MAX_ATTACHMENTS) {
            "Maximum ${AppConstants.CLAIMS.MAX_ATTACHMENTS} attachments allowed"
        }
        require(serviceDate <= Date()) {
            "Service date cannot be in the future"
        }
    }

    /**
     * Enhanced check if claim can be edited based on status and security context
     * @param userId ID of user attempting to edit
     * @return Boolean indicating if claim is editable
     */
    fun isEditable(userId: String): Boolean {
        return when {
            status in setOf(ClaimStatus.APPROVED, ClaimStatus.REJECTED) -> false
            securityContext["locked_by"]?.let { it != userId } == true -> false
            else -> true
        }
    }

    /**
     * Enhanced calculation of expected reimbursement with policy rules
     * @return Double representing expected reimbursement amount
     */
    fun calculateReimbursement(): Double {
        val baseAmount = amount * (metadata.coveragePercentage / 100)
        return maxOf(0.0, baseAmount - metadata.deductibleApplied)
    }

    /**
     * Adds secure audit trail entry for claim modifications
     * @param userId ID of user performing action
     * @param action Description of action performed
     * @param reason Reason for modification
     */
    fun addAuditEntry(userId: String, action: String, reason: String) {
        val entry = AuditEntry(
            userId = userId,
            action = action,
            details = reason,
            ipAddress = "INTERNAL", // Would be populated from actual request
            deviceInfo = "SYSTEM"   // Would be populated from actual request
        )
        auditTrail + entry
    }

    companion object {
        /**
         * Retention period in years for claim records per HIPAA requirements
         */
        const val RETENTION_PERIOD_YEARS = AppConstants.CLAIMS.RETENTION_PERIOD_YEARS
    }
}